src/darwin/Framework/CHIP/MTRDeviceControllerDataStore.mm

```diff
@@ -0,0 +1,238 @@
+/**
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "MTRDeviceControllerDataStore.h"
+#import "MTRLogging_Internal.h"
+
+// TODO: FIXME: Figure out whether these are good key strings.
+static NSString * sResumptionNodeListKey = @"caseResumptionNodeList";
+static NSString * sLastUsedNOCKey = @"lastUsedControllerNOC";
+
+static NSString * ResumptionByNodeIDKey(NSNumber * nodeID)
+{
+    return [NSString stringWithFormat:@"caseResumptionByNodeID/%llx", nodeID.unsignedLongLongValue];
+}
+
+static NSString * ResumptionByResumptionIDKey(NSData * resumptionID)
+{
+    return
+        [NSString stringWithFormat:@"caseResumptionByResumptionID/%s", [resumptionID base64EncodedStringWithOptions:0].UTF8String];
+}
+
+@implementation MTRDeviceControllerDataStore {
+    id<MTRDeviceControllerStorageDelegate> _storageDelegate;
+    dispatch_queue_t _storageDelegateQueue;
+    MTRDeviceController * _controller;
+    // Array of nodes with resumption info, oldest-stored first.
+    NSMutableArray<NSNumber *> * _nodesWithResumptionInfo;
+}
+
+- (instancetype)initWithController:(MTRDeviceController *)controller
+                   storageDelegate:(id<MTRDeviceControllerStorageDelegate>)storageDelegate
+              storageDelegateQueue:(dispatch_queue_t)storageDelegateQueue
+{
+    if (!(self = [super init])) {
+        return nil;
+    }
+
+    _controller = controller;
+    _storageDelegate = storageDelegate;
```

Comment:
Check to see if nil passed in?

Reply:
The API says this is not nullable.  Should we still be guarding against it being nil? 

Concretely, the one caller we have does not pass nil.

Reply:
I'm fine with it, I just prefer to litter nil checks everywhere.